src/Adapter/Language/LanguageCopier.php

```diff
@@ -0,0 +1,187 @@
+ */
+final class LanguageCopier implements LanguageCopierInterface
+{
+    /**
+     * @var LanguageDataProvider
+     */
+    private $languageDataProvider;
+
+    /**
+     * @var array
+     */
+    private $errors = [];
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @param LanguageDataProvider $languageDataProvider
+     * @param TranslatorInterface $translator
+     */
+    public function __construct(
+        LanguageDataProvider $languageDataProvider,
+        TranslatorInterface $translator
+    ) {
+        $this->languageDataProvider = $languageDataProvider;
+        $this->translator = $translator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function copyLanguage(LanguageCopierConfigInterface $config)
+    {
+        $languageFiles = $this->languageDataProvider->getFilesList(
+            $config->getLanguageFrom(),
+            $config->getThemeFrom(),
+            $config->getLanguageTo(),
+            $config->getThemeTo(),
+            true
+        );
+
+        $fs = new Filesystem();
+
+        foreach ($languageFiles as $source => $destination) {
+            try {
+                $fs->mkdir(dirname($destination));
+            } catch (IOExceptionInterface $exception) {
+                $this->errors[] = $this->translator->trans(
+                    'Cannot create the folder "%folder%". Please check your directory writing permissions.',
+                    [
+                        '%folder%' => $destination,
+                    ],
+                    'Admin.International.Notification'
+                );
+                continue;
+            }
+
+            try {
+                $fs->copy($source, $destination);
+            } catch (IOExceptionInterface $exception) {
+                $this->errors[] = $this->translator->trans(
+                    'Impossible to copy "%source%" to "%dest%".',
+                    [
+                        '%source%' => $source,
+                        '%dest%' => $destination,
+                    ],
+                    'Admin.International.Notification'
+                );
+                continue;
+            }
+
+            if ($this->isModuleContext($source, $destination, $config->getLanguageFrom())) {
+                $changedModuleTranslationKeys = $this->changeModulesTranslationKeys(
+                    $destination,
+                    $config->getThemeFrom(),
+                    $config->getThemeTo()
+                );
+
+                if (!$changedModuleTranslationKeys) {
+                    $this->errors[] = $this->translator->trans(
+                        'Impossible to translate "%dest%".',
+                        [
+                            '%dest%' => $destination,
+                        ],
+                        'Admin.International.Notification'
+                    );
+                }
+            }
+        }
+
+        if (!empty($this->errors)) {
```

Comment:
hmm, i think you should clear `$this->errors` everytime `copyLanguage()` is called, since errors from previous call may still exist, for example:
```php
$languageCopier->copyLanguage($config); // produces errors
$errors = $languageCopier->getErrors(); // get errors from previous copying
$languageCopier->copyLanguage($config); // produces errors again
$errors = $languageCopier->getErrors(); // get errors both previous copies
```

am i right?

on the other hand, why not returing errors from `copyLanguage()`? i think it would be best not to store any state in service, wdyt @mickaelandrieu ?